File: CSC330/poker/Julia/Deck.jl
# Deck Julia Class
using Random
include("Card.jl")

mutable struct Deck
	cards::Vector{Card}
	
	function Deck()
		suits = ["S","H","C","D"]
		ranks = ["A","2","3","4","5","6","7","8","9","10","J","Q","K"]
		deck = [Card(rank, suit) for rank in ranks, suit in suits]
		deck = vec(deck)
		# Shuffle the deck as soon as it is created. Eliminates the need for two functions
		shuffle!(deck)
		new(deck)
	end
end

function deal(deck::Deck)
	return pop!(deck.cards)
end

function printDeck(deck::Deck)
	i::Int = 0
	for card in reverse(deck.cards)
		print(card, " ")
		i += 1
		if ( i % 13 == 0 )
			println()
		end
	end
	println()
end

function deal_card!(deck::Deck)
    return pop!(deck.cards)
end

File: CSC330/poker/Python/deck.py
# Deck Class - Python

from card import Card
import random

class Deck:
    
    def __init__(self):
        self.cardDeck = []
        
        suits = ['h', 'c', 's', 'd']
        ranks = ['2','3','4','5','6','7','8','9','10','J','Q','K','A']
        for suit in suits:
            for rank in ranks:
                self.cardDeck.append(Card(suit, rank))
    

    def shuffle(self):
        random.shuffle(self.cardDeck)
    
    def dealCard(self):
        return self.cardDeck.pop()
    
    def printDeck(self):
        print("Deck\n----")
        for card in self.cardDeck:
            print(f"{card.rank}{card.suit.upper()}",end="  ")
        
        print("\n----")
    
    def __str__(self):
        return ", ".join([f"{card.rank}{card.suit.upper()}" for card in self.cardDeck])
